spock-core/src/main/java/org/spockframework/runtime/Scheduler.java

```diff
@@ -0,0 +1,67 @@
+package org.spockframework.runtime;
+
+import org.junit.runners.model.RunnerScheduler;
+
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class Scheduler {
+  private final RunnerScheduler runnerScheduler;
+  private final boolean forceSequential;
+  private final AtomicInteger scheduledTasksCount = new AtomicInteger(0);
+  private final Semaphore concurrentJobsSemaphore = new Semaphore(0);
+  private State state = State.SCHEDULING;
```

Comment:
https://stackoverflow.com/questions/7855700/why-is-volatile-used-in-this-example-of-double-checked-locking